### PR TITLE
Add BytesCountingFilterInputStream for counting bytes read

### PR DESCRIPTION
**Pull request by ywangd**
See title

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStream.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.io.stream;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class BytesCountingFilterInputStream extends FilterInputStream {
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStream.java:15]: nit Seeing there's a `CountingStreamOutput`, maybe it'd be nice if this was renamed `CountingFilterInputStream`.

File: server/src/test/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStreamTests.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        final int bytesBefore = in.getBytesRead();
+        final byte[] remainingBytes = in.readAllBytes();
+        assertThat(in.getBytesRead(), equalTo(bytesBefore + remainingBytes.length));
+        assertThat(in.getBytesRead(), equalTo(input.length));
+
+        // Read beyond available data has no effect
+        in.read();
+        final byte[] bytes = new byte[between(20, 30)];
```
Comment (kingherc) [inline server/src/test/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStreamTests.java:88]: nit Maybe you can set it to zeroes, and assert it's still equal to zeroes at the end of the test.

**Review by kingherc — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStream.java (hunk 1 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStream.java:15]: Sounds good. Renamed as suggested.

File: server/src/test/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStreamTests.java (hunk 2 above)
Comment (ywangd) [reply server/src/test/java/org/elasticsearch/common/io/stream/BytesCountingFilterInputStreamTests.java]: Thanks. Pushed 0436d246f4955181a66352998be82200e64a801e

**Comment by ywangd** [comment]
@elasticmachine update branch

**Comment by ywangd** [comment]
@elasticmachine update branch

**Pull request merged**